dependency.go

```diff
@@ -0,0 +1,171 @@
+	if slices.Index(curNetworks, network.Name) == -1 {
+		err = network.provider.client.NetworkConnect(ctx, network.ID, container.GetContainerID(), nil)
+		if err != nil {
+			return "", fmt.Errorf("%w: could not connect dependency container to network", err)
+		}
+	}
+
+	networkAlias, err := container.NetworkAliases(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	aliases := networkAlias[network.Name]
+	if len(aliases) == 0 {
+		return "", errors.New("could not retrieve network alias for dependency container")
+	}
+	return aliases[0], nil
+}
+
+func cleanupDependencyNetwork(ctx context.Context, dependencies map[Container]bool, network *DockerNetwork) error {
+	if network == nil {
+		return nil
+	}
+
+	for dependency, keepAlive := range dependencies {
+		err := network.provider.client.NetworkDisconnect(ctx, network.ID, dependency.GetContainerID(), true)
+		if err != nil {
+			return err
+		}
+
+		if !keepAlive {
+			if err := dependency.Terminate(ctx); err != nil {
+				return err
+			}
+		}
+
+	}
+	defer network.provider.Close()
+	return network.Remove(ctx)
+}
+
+var defaultDependencyHook = func(dockerInput *container.Config) ContainerLifecycleHooks {
+	var depNetwork *DockerNetwork
+	depContainers := make(map[Container]bool)
+	return ContainerLifecycleHooks{
+		PreCreates: []ContainerRequestHook{
+			func(ctx context.Context, req ContainerRequest) (err error) {
+				if len(req.DependsOn) == 0 {
+					return nil
+				}
+				defer func() {
+					if err != nil {
+						// clean up dependencies that were created if an error occurred.
+						cleanupErr := cleanupDependencyNetwork(ctx, depContainers, depNetwork)
+						if cleanupErr != nil {
+							Logger.Printf("Could not cleanup dependencies after an error occured: %v", cleanupErr)
+						}
+					}
+				}()
+
+				net, err := GenericNetwork(ctx, GenericNetworkRequest{
+					NetworkRequest: NetworkRequest{
+						Driver:   Bridge,
+						Labels:   GenericLabels(),
+						Name:     fmt.Sprintf("testcontainer-dependency-%v", uuid.NewString()),
+						Internal: false,
+					},
+				})
+				depNetwork = net.(*DockerNetwork)
+				if err != nil {
+					return err
+				}
+
+				for _, dep := range req.DependsOn {
+					if dep.EnvKey == "" {
+						return errors.New("cannot create dependency with empty environment key.")
+					}
+					container, err := dep.StartDependency(ctx, depNetwork.Name)
+					if err != nil {
+						return err
+					}
+					depContainers[container] = dep.KeepAlive
+					name, err := resolveDNSName(ctx, container, depNetwork)
+					if err != nil {
+						return err
+					}
+					dockerInput.Env = append(dockerInput.Env, dep.EnvKey+"="+name)
+				}
+				return nil
+			},
+		},
+		PostCreates: []ContainerHook{
+			func(ctx context.Context, container Container) error {
+				if depNetwork == nil {
+					return nil
+				}
+				err := depNetwork.provider.client.NetworkConnect(ctx, depNetwork.ID, container.GetContainerID(), nil)
+				defer depNetwork.provider.Close()
+				return err
```

Comment:
suggestion: check and wrap
```suggestion
				defer depNetwork.provider.Close()
				if err := depNetwork.provider.client.NetworkConnect(ctx, depNetwork.ID, container.GetContainerID(), nil); err != nil {
					return fmt.Errorf("connect network: %w", err)
				}

				return nil
```